8225763: Inflater and Deflater should implement AutoCloseable

jaikiran opened the pull request:
Can I please get a review of this enhancement which proposes to enhance `java.util.zip.Deflater/Inflater` classes to now implement `AutoCloseable`?

The actual work for this was done a few years back when we discussed the proposed approaches and then I raised a RFR. At that time I couldn't take this to completion. The current changes in this PR involve the implementation that was discussed at that time and also have implemented the review suggestions from that time. Here are those previous discussions and reviews:

https://mail.openjdk.org/pipermail/core-libs-dev/2019-June/061079.html
https://mail.openjdk.org/pipermail/core-libs-dev/2019-July/061177.html
https://mail.openjdk.org/pipermail/core-libs-dev/2019-July/061229.html

To summarize those discussions, we had concluded that:
- `Deflater` and `Inflater` will implement the `AutoCloseable` interface
-  In the `close()` implementation we will invoke the `end()` method (`end()` can be potentially overridden by subclasses).
- `close()` will be specified and implemented to be idempotent. Calling `close()` a second time or more will be a no-op.
- Calling `end()` and then `close()`, although uncommon, will also support idempotency and that `close()` call will be a no-op.
- However, calling `close()` and then `end()` will not guarantee idempotency and depending on the implementing subclass, the `end()` may throw an exception.

New tests have been included as part of these changes and they continue to pass along with existing tests in tier1, tier2 and tier3. When I had originally added these new tests, I hadn't used junit. I can convert them to junit if that's preferable.

I'll file a CSR shortly.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8343768](https://bugs.openjdk.org/browse/JDK-8343768) to be approved

### Issues
 * [JDK-8225763](https://bugs.openjdk.org/browse/JDK-8225763): Inflater and Deflater should implement AutoCloseable (**Enhancement** - P4)
 * [JDK-8343768](https://bugs.openjdk.org/browse/JDK-8343768): Inflater and Deflater should implement AutoCloseable (**CSR**)


### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**) Review applies to [19f2db96](https://git.openjdk.org/jdk/pull/19675/files/19f2db960d8da123fc0b2225f70986c1c9d2ffb4)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19675/head:pull/19675` \
`$ git checkout pull/19675`

Update a local copy of the PR: \
`$ git checkout pull/19675` \
`$ git pull https://git.openjdk.org/jdk.git pull/19675/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19675`

View PR using the GUI difftool: \
`$ git pr show -t 19675`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19675.diff">https://git.openjdk.org/jdk/pull/19675.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19675#issuecomment-2162705103)
</details>

jaikiran commented:
/csr needed

liach commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -872,14 +872,41 @@ public void reset() {
+    @Override
+    public void close() {
+        synchronized (zsRef) {
+            // check if already closed
```
Should we comment `// in case subclasses override the closed check in end()` instead? Otherwise the duplicate comments and checks are confusing.

jaikiran replied:
Hello Chen, we want the close() to be idempotent irrespective of whether or not end() is overridden. That check in close() and the code comment on that line is to indicate that. If the comment is adding to confusion, I can remove it - I don't think it's a must to have it to understand the check.

liach replied:
I mean, we have updated the default implementation of end() to be idempotent. It now just appears confusing to readers that close() **seems to** perform redundant checks for idempotence until they realize subclasses can override end() to make it non-idempotent.

My suggestion is that our comment should say why we have this check here even though there's already an identical check in end().

jaikiran replied:
> My suggestion is that our comment should say why we have this check here even though there's already an identical check in end().

I now see what you mean. I'll update this appropriately tomorrow.

liach commented on src/java.base/share/classes/java/util/zip/Inflater.java:
```diff
@@ -691,18 +691,44 @@ public void reset() {
+    @Override
+    public void close() {
+        synchronized (zsRef) {
+            // check if already closed
```
Same remark, we should mention subclass override close risk

liach commented on test/jdk/java/util/zip/DeflaterClose.java:
```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2019, Oracle and/or its affiliates. All rights reserved.
```
```suggestion
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
Same for `InflaterClose.java`.

jaikiran replied:
Hello Chen, these tests were actually introduced in a RFR back in 2019, so the original year is correct. I've now converted these tests to junit and updated the PR and updated the copyright years as well.

stuart-marks commented:
Hi Jai, thanks for picking this up. Several points of discussion here. Nothing fatal, but as I said in 2019 :-) just a few wrinkles to iron out. I looked mainly at Deflater, but I think the issues also all apply to Inflater too.

1) class specification

I think the class specification needs to be clearer about the positioning of end() and close(). The end() method has done the real work of "closing" since the classes were introduced. We're retrofitting them to to have a close() method that's essentially just a synonym for end(), and it's still perfectly legal for clients and subclasses to call end() instead of close(). I think we need to say that close() is present mainly to support AutoCloseable and try-with-resources.

2) end() specification

I don't think that "may throw an exception" is strong enough. Also, on these classes (not subclasses) end() is idemopotent, isn't it? If so, this should be specified. The end() specs need to say something like, closes and releases resources, etc., henceforth operations will throw an exception (which? -- see below) but that subsequent calls to end() do nothing.

3) Which exceptions on closed Deflater/Inflater?

Oddly, it doesn't seem to be specified anywhere what exceptions are thrown if operations are attempted on a closed object. And the implementation actually throws NPE??!? Ideally we should specify what exception is thrown in this circumstance, but NPE seems wrong. Maybe we want to change the behavior to throw a different exception type. IllegalStateException seems like a reasonable candidate.

4) close() specification

The 2019 discussion kind of assumed that we want close() to be idempotent. I'm not sure this is necessary. Alan said the implementation might need to be "heroic" but Peter Levart's arrangement (carried through here, apparently) isn't terribly bad. And @liach suggests more comments on this code, quite reasonably... but the comment begs the question, why do we need to avoid calling end() more than once? I'm rethinking the need for close() to be idempotent. Maybe close() should simply call end(). It would simplify the implementation and the specs -- the implSpec would simply say that it calls the end() method.

This would only be a problem if a) some arrangement of TWR ends up calling close() twice, which apparently it doesn't; b) there is a subclass; and c) the subclass's end() method is not idempotent. As far as I can tell we haven't found any example of any of these. It thus seems to me that having extra logic in close() to avoid calling end() more than once is solving a non-problem. It also fits more closely to the model of "close() is simply a synonym for end()".

jaikiran commented:
Please keep open, I'm running some experiments for the proposed alternative and I'll update this PR shortly.

jaikiran commented:
/open

jaikiran commented:
Thank you Stuart for your inputs.

> class specification
>
> I think the class specification needs to be clearer about the positioning of end() and close(). The end() method has done the real work of "closing" since the classes were introduced. We're retrofitting them to to have a close() method that's essentially just a synonym for end(), and it's still perfectly legal for clients and subclasses to call end() instead of close(). I think we need to say that close() is present mainly to support AutoCloseable and try-with-resources.

I've now updated the class level javadoc of both Inflater and Deflater to be more explicit on the purpose of close() method and how it relates to the end() method.


> end() specification
>
> I don't think that "may throw an exception" is strong enough. Also, on these classes (not subclasses) end() is idemopotent, isn't it? If so, this should be specified. The end() specs need to say something like, closes and releases resources, etc., henceforth operations will throw an exception (which? -- see below) but that subsequent calls to end() do nothing.

I've updated both the end() and close() method javadocs. It now states that end() is idempotent and at the same time states that calling several other methods on a closed Inflater/Deflater will throw a specific exception (noted below). I copied over the sentence which states "Several other methods defined..." from an existing Java SE API.


> Which exceptions on closed Deflater/Inflater?
>
> Oddly, it doesn't seem to be specified anywhere what exceptions are thrown if operations are attempted on a closed object. And the implementation actually throws NPE??!? Ideally we should specify what exception is thrown in this circumstance, but NPE seems wrong. Maybe we want to change the behavior to throw a different exception type. IllegalStateException seems like a reasonable candidate.

Yes, it is indeed odd that it's throwing a `NullPointerException` when ensuring the Inflater/Deflater is open. I have changed that part to throw an `IllegalStateException` and make a mention of this exception in the end() method specification. Clearly, changing `NullPointerException` to `IllegalStateException` is a change in behaviour of public APIs and ideally should be done separately. But, given the context that it gets thrown only from an instance that's already closed (and thus of no real use), I think the chances of this change severly impacting applications is very limited. So, in my opinion, it's perhaps OK to include this change of exception type as part of this change and have it recorded through the same CSR that we use for this current enhancement. Do let me know if we should do that change separately.

> close() specification
>
> The 2019 discussion kind of assumed that we want close() to be idempotent. I'm not sure this is necessary. Alan said the implementation might need to be "heroic" but Peter Levart's arrangement (carried through here, apparently) isn't terribly bad. And @liach suggests more comments on this code, quite reasonably... but the comment begs the question, why do we need to avoid calling end() more than once? I'm rethinking the need for close() to be idempotent. Maybe close() should simply call end(). It would simplify the implementation and the specs -- the implSpec would simply say that it calls the end() method.

I ran various corpus experiments. Based on that what I have noticed is that, in that corpus, there are extremely few (less than 10) classes which override end() of either the Inflater or Deflater class. Furthermore, all these implementations have implemented their end() in a manner that calling it more than once will act as a no-op for subsequent invocations. So I think, like you note, we should just specify that close() calls end() and stop trying to prevent end() being called more than once. Given this, I've now updated the PR to do just that and also updated the new tests to match this specification updates.

tier testing is currently in progress for this current state of the PR. If the current state of the PR looks reasonable, then I'll focus on the CSR text for these changes.

liach commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -868,23 +870,44 @@ public void reset() {
+     * @since 24
+     */
+    @Override
+    public void close() {
```
I agree with marking these close methods as final as long as the previous corpus search did not find any user-defined close method in subclasses.

jaikiran replied:
Hello Chen, corpus analysis shows that there are no subclasses of Inflater/Deflater which have a `close()` method.

jaikiran replied:
I have updated the PR to follow Roger's suggestion of marking this new `close()` method on `Inflater` and `Deflater` classes as `final`.

LanceAndersen commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -60,8 +60,10 @@
+ * The {@code Deflater} cleans up its resources in its {@link #end()} method. To facilitate
+ * usage with try-with-resources, this class implements {@link AutoCloseable}. The
+ * {@link #close()} method of this class calls {@code end()} to clean up its
+ * resources. Subclasses are responsible for the cleanup of resources
```
As part of the clarification, do we need to state the verbiage above regarding try-with-resources specifically?

I did a quick scan of some of the other classes implementing AutoClosable and did not see many cases of that in the documentation most likely due to that behavior is clear in the documentation for AutoClosable

jaikiran replied:
Stuart in one of his review comments in this PR had noted that:

> I think the class specification needs to be clearer about the positioning of end() and close(). The end() method has done the real work of "closing" since the classes were introduced. We're retrofitting them to to have a close() method that's essentially just a synonym for end(), and it's still perfectly legal for clients and subclasses to call end() instead of close(). I think we need to say that close() is present mainly to support AutoCloseable and try-with-resources.

The current wording of that class level javadoc is my attempt to implement that suggestion. Do you think we can word it differently and yet convey the purpose of close()?

LanceAndersen replied:
Hi Jai,

I am not convinced we need to call this out specifically at the class level given  the classes implements AutoClosable and close indicates it calls end.  We will also  want to add a Release Note which can help further clarify this

Now that being said, what I would suggest instead is a simple code example in the class header which using try-with-resources would be more effective/helpful.

jaikiran replied:
Hello Stuart, I've now updated the PR to follow your recommendations for the class level javadoc of the `Deflater`. Does this look OK?

Like you note, I'll update the `Inflater` similarly once we settle on the text for the `Deflater`.

LanceAndersen commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -868,23 +870,44 @@ public void reset() {
+     * <p>
+     * If this method is invoked multiple times, the subsequent calls are treated as a no-op.
+     * Several other methods defined by this class will throw an {@link IllegalStateException}
+     * if invoked on a closed {@code Deflater}.
```
All of the methods which call ensureOpen should either document that an IllegalStateException may be thrown or you could add it to the header of the class.

jaikiran replied:
Hello Lance, you are right - I forgot to do that when I changed it to throw IllegalStateException from the previous NullPointerException. I've now updated the PR to document this on individual methods in the Inflater/Deflater which throw this exception.

LanceAndersen replied:
Thank you.  It was an oversight that the original javadoc did not call out the NPE, but now is our chance to get it right with a more reasonable exception.

stuart-marks replied:
Overall I agree with defining that IllegalStateException is thrown after the Deflater is closed, and adding documentation of this to all the methods that throw it.

LanceAndersen replied:
I am no sure the last sentence, “several…” is needed here given we now document the IAE is thrown

LanceAndersen reviewed (commented):
HI Jai,

Thanks for tackling this and sticking with it.  I have few comments below which could be considered for both Deflater/Inflater

Also, extra credit for:

-  converting the CloseinflaterDeflaterTest to use junit from testNG
- converting TotalInOut to use junit from the older test structure

jaikiran commented on src/java.base/share/classes/java/util/zip/Inflater.java:
```diff
@@ -684,25 +686,45 @@ public void reset() {
+     * @since 24
+     */
+    @Override
+    public void close() {
```
Done.

jaikiran replied:
Hello Alan, 

> If it goes ahead as reviewed, then we'll to do some outreach and be prepared to back out this change in the event that breakage is reported before GA.

I agree that if the `final close()` causes issues, then backing this change out and reconsidering a non-final `close()` would be the right thing to do. Once/if this gets integrated into 24, I will work with the outreach team to have this change noted in their EA mails.

AlanBateman replied:
Adding a final close method is both source and binary incompatible so I don't think the CSR can be approved without more justification. So I think the issues of introducing a non-final close should be examined again.

jaikiran replied:
Hello Alan, I've now reverted the change which had marked this new `close()` method on `Inflater` and `Deflater` as `final`. With this update to this PR, this method is no longer `final`. Given this change, one additional sentence has been added to the Inflater/Deflater class level javadoc, in `@apiNote` section providing guidance to the subclasses on which method to override for cleaning up resources they hold:

>
> Subclasses should override {@linkplain #end()} to clean up the resources acquired by the subclass.
>

No other changes have been done.

Requesting your and others' inputs if this updated change looks reasonable.

stuart-marks commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -868,23 +886,44 @@ public void reset() {
+     * Closes and releases the resources held by this {@code Deflater}
+     * and discards any unprocessed input.
+     * <p>
+     * If this method is invoked multiple times, the subsequent calls are treated as a no-op.
```
I would say,

> If this method is invoked multiple times, the second and subsequent calls do nothing.

jaikiran replied:
Done. I've updated the end() methods javadoc of both `Deflater` and `Inflater` to use this recommended text.

stuart-marks commented on src/java.base/share/classes/java/util/zip/Inflater.java:
```diff
@@ -59,8 +59,10 @@
+ * The {@code Inflater} cleans up its resources in its {@link #end()} method. To facilitate
+ * usage with try-with-resources, this class implements {@link AutoCloseable}. The
+ * {@link #close()} method of this class calls {@code end()} to clean up its
+ * resources. Subclasses are responsible for the cleanup of resources
```
Once we've agreed on the arrangement in the Deflater class, the corresponding changes should be made to Inflater, mutatis mutandis.

RogerRiggs reviewed (approved):
Thanks for the improvements.

stuart-marks reviewed (approved):
Marking as reviewed. Don't forget to update the Inflater class with similar wording changes as Deflater. Then copy the spec diffs into the CSR and I'll review it too.

jaikiran commented:
Thank you Stuart for the review. I've now updated the class level doc of `Inflater` to match the one in `Deflater`. The CSR Specification section has now been updated to use this new text.

LanceAndersen reviewed (approved):
Morning Jai

thank you for the latest updates looks good overall. I do have one suggested change which I would consider making.

jaikiran commented:
Hello Lance,

> I am no sure the last sentence, “several…” is needed here given we now document the IAE is thrown

I think what you suggest is reasonable. I've updated the javadoc of `end()` methods of both Inflater and Deflater to remove that sentence. The CSR Specification diff has been updated too.

RogerRiggs commented on src/java.base/share/classes/java/util/zip/Inflater.java:
```diff
@@ -49,23 +49,24 @@
- * and decompression of a string using {@code Deflater} and
- * {@code Inflater}.
- * {@snippet id="compdecomp" lang="java" class="Snippets" region="DeflaterInflaterExample"}
+ * To release resources used by the {@code Inflater}, applications must call the
+ * {@link #end()} method. After {@code end()} has been called, subsequent calls
+ * to several methods of the {@code Inflater} will throw an {@link IllegalStateException}.
```
Since `close()` is not mentioned here, there may be some come confusion about whether `end()` still needs to be called withing T-W-R.

jaikiran replied:
I think that's a good point. I've now updated the PR to reword this part of the text both in `Inflater` and `Deflater`. Does the change look OK?

RogerRiggs commented on test/jdk/java/util/zip/DeflaterClose.java:
```diff
@@ -32,15 +32,15 @@
 
     /**
-     * Closes the Deflater multiple times and then expects end() to be called that
+     * Closes the Deflater multiple times and then expects close() and end() to be called that
```
Should this be "close() *or* end()"?

jaikiran replied:
Hello Roger, calling `close()` on Inflater/Deflater will just call `end()` on the instance. So I believe "close() and end()" in the test comment here is correct. I now pushed a very minor change just to make it more clear. Let me know if you think we should improve the comment further.

RogerRiggs commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -799,6 +812,8 @@ public int getAdler() {
      * and therefore cannot return the correct value when it is greater
      * than {@link Integer#MAX_VALUE}.
      *
+     * @throws IllegalStateException if the Deflater is closed
```
\@throws should follow \@return;  (Style notes in https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html).

jaikiran replied:
Done - I've updated the PR to fix this and the other places that were introduced in this PR to follow that style guide.

RogerRiggs commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -829,6 +845,8 @@ public long getBytesRead() {
      * and therefore cannot return the correct value when it is greater
      * than {@link Integer#MAX_VALUE}.
      *
+     * @throws IllegalStateException if the Deflater is closed
```
Ditto `@throws` throws should follow `@return`

RogerRiggs commented on src/java.base/share/classes/java/util/zip/Inflater.java:
```diff
@@ -646,6 +660,8 @@ public long getBytesRead() {
      * and therefore cannot return the correct value when it is greater
      * than {@link Integer#MAX_VALUE}.
      *
+     * @throws IllegalStateException if the Inflater is closed
```
Ditto ` @throws` should follow `@return`

LanceAndersen reviewed (commented):
Overall looks good Jai.

Roger is right on moving the @throws IAE further down

RogerRiggs reviewed (approved):
Thanks for the updates.

jaikiran commented:
Thank you all for the reviews and the inputs. The CSR has been approved. I'll go ahead and integrate this shortly.

AlanBateman commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -49,29 +49,33 @@
+ * To release resources used by the {@code Deflater}, applications must close the
+ * {@code Deflater} by calling either the {@link #end()} or the {@link #close()} method.
+ * After the {@code Deflater} has been closed, subsequent calls to several methods
+ * of the {@code Deflater} will throw an {@link IllegalStateException}.
```
This paragraph uses the definite article but there isn't a specific Deflater to speak of, and it's not a singleton. The first sentence of this paragraph might be better if re-worded "To release the resources used a Deflater, an application must close it by invoking its end() or close() method".

jaikiran replied:
I have reworded the sentence to follow your input. Addtionally, I've removed the second sentence, since as you note, relevant methods on the Inflater/Deflater (as part of this PR) already have been updated to state that they throw an `IllegalStateException` if the Inflater/Deflater is already closed.

AlanBateman commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -868,23 +886,42 @@ public void reset() {
+     * Closes and releases the resources held by this {@code Deflater}
+     * and discards any unprocessed input.
+     * <p>
+     * If this method is invoked multiple times, the second and subsequent calls do nothing.
```
I think this could be clearer if you replace this sentence with "If the Deflater is already closed then invoking this method has no effect."

jaikiran replied:
Done.

AlanBateman commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -49,29 +49,33 @@
+ * This class implements {@link AutoCloseable} to facilitate its usage with
+ * {@code try}-with-resources statement. The {@linkplain Deflater#close() close() method} simply
+ * calls {@code end()}. Subclasses should override {@linkplain #end()} to clean up the
+ * resources acquired by the subclass.
```
I wonder if the note for subclasses should move to the end method as implSpec, it looks out of place in the class API note.

liach replied:
I believe an `@implSpec` only restricts the implementation of this immediate method, such as those seen on default methods, and has no impact on the overrides. If we need to specify something for overrides to abide to, I believe they usually stay in the API specification itself.

jaikiran replied:
> I wonder if the note for subclasses should move to the end method as implSpec, it looks out of place in the class API note.

Stuart, in his inputs here https://github.com/openjdk/jdk/pull/19675#issuecomment-2172433871 and here https://github.com/openjdk/jdk/pull/19675#discussion_r1848945718 had noted that it would be useful to add this guidance to subclasses. But I see that he wasn't fully convinced whether it must be a apiNote or something else. Given what you note about moving this detail for subclasses to a `@implSpec` on the end() method, I have now updated the PR accordingly.

jaikiran replied:
Hello Chen,
> I believe an @implSpec only restricts the implementation of this immediate method, such as those seen on default methods, and has no impact on the overrides.

Based on the details about these tags here https://openjdk.org/jeps/8068562, I think the `@implSpec` is the most relevant one for this text:

> Interface implementors or class subclassers use the information here in order to decide whether it is sensible or necessary to override a particular method

jaikiran commented on src/java.base/share/classes/java/util/zip/Deflater.java:
```diff
@@ -49,29 +49,33 @@
  * <p>
  * This class deflates sequences of bytes into ZLIB compressed data format.
- * The input byte sequence is provided in either byte array or byte buffer,
+ * The input byte sequence is provided in either byte array or {@link ByteBuffer},
```
Fixed.

AlanBateman reviewed (approved):
Thanks for the updates, looks good now.

LanceAndersen reviewed (approved):
Latest Updates look good Jai

jaikiran commented:
Thank you all for the reviews. tier1, tier2 and tier3 tests pass with this change. I'll now go ahead and integrate this.

jaikiran commented:
/integrate
